### PR TITLE
hotfix: return error instead of log at `FromMapAndOption`

### PR DESCRIPTION
**Pull request by chansuke**
Make [FromMapAndOption](https://github.com/kubernetes-sigs/kustomize/blob/master/api/resource/factory.go#L64-L65) to return error.

### CODE REVIEW

File: api/internal/plugins/utils/utils_test.go [hunk 1]
```diff
@@ -89,7 +89,8 @@ func TestUpdateResourceOptions(t *testing.T) {
 		name := fmt.Sprintf("test%d", i)
 		err := in.Append(makeConfigMap(rf, name, c.behavior, c.hashValue))
 		require.NoError(t, err)
-		err = expected.Append(makeConfigMapOptions(rf, name, c.behavior, !c.needsHash))
+		config := makeConfigMap(rf, name, c.behavior, c.hashValue)
```
Comment (DiptoChakrabarty) [inline api/internal/plugins/utils/utils_test.go:92]: it seems you have replaced the `makeConfigMapOptions` with `makeConfigMap` function which is causing a linter error.

Additionally this is eliminating the test for  `makeConfigMapOptions` function.

Could you fix this.

**Comment by DiptoChakrabarty** [comment]
also seems like the tests are failing can you go ahead and fix them

File: api/internal/plugins/utils/utils_test.go (hunk 1 above)
Comment (chansuke) [reply api/internal/plugins/utils/utils_test.go:92]: @DiptoChakrabarty Thank you so much for your review.I will fix this.

File: api/internal/plugins/utils/utils_test.go [hunk 2]
```diff
@@ -89,7 +89,11 @@ func TestUpdateResourceOptions(t *testing.T) {
 		name := fmt.Sprintf("test%d", i)
 		err := in.Append(makeConfigMap(rf, name, c.behavior, c.hashValue))
 		require.NoError(t, err)
-		err = expected.Append(makeConfigMapOptions(rf, name, c.behavior, !c.needsHash))
+		config, err := makeConfigMapOptions(rf, name, c.behavior, !c.needsHash)
+		if err != nil {
+			t.Errorf("unexpected error: %v", err)
```
Comment (koba1t) [inline api/internal/plugins/utils/utils_test.go:94]: I think the `unexpected error` message is confusing.
Could you add any meaningful message?

File: api/resmap/reswrangler_test.go [hunk 3]
```diff
@@ -836,11 +837,15 @@ func makeMap1() ResMap {
 			},
 		}, &types.GeneratorArgs{
 			Behavior: "create",
-		}))
+		})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
```
Comment (koba1t) [inline api/resmap/reswrangler_test.go:842]: same here

File: api/resmap/reswrangler_test.go [hunk 4]
```diff
@@ -854,14 +859,19 @@ func makeMap2(b types.GenerationBehavior) ResMap {
 			},
 		}, &types.GeneratorArgs{
 			Behavior: b.String(),
-		}))
+		})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
```
Comment (koba1t) [inline api/resmap/reswrangler_test.go:864]: same too

File: api/resmap/reswrangler_test.go [hunk 5]
```diff
@@ -874,24 +884,28 @@ func TestAbsorbAll(t *testing.T) {
 		},
 		&types.GeneratorArgs{
 			Behavior: "create",
-		}))
-	w := makeMap1()
-	assert.NoError(t, w.AbsorbAll(makeMap2(types.BehaviorMerge)))
+		})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
```
Comment (koba1t) [inline api/resmap/reswrangler_test.go:889]: same too

File: api/resmap/reswrangler_test.go [hunk 6]
```diff
@@ -874,24 +884,28 @@ func TestAbsorbAll(t *testing.T) {
...
 	assert.NoError(t, w.AbsorbAll(w2))
 	w2.RemoveBuildAnnotations()
 	assert.NoError(t, w2.ErrorIfNotEqualLists(w))
-	w = makeMap1()
-	w2 = makeMap2(types.BehaviorUnspecified)
-	err := w.AbsorbAll(w2)
+	w = makeMap1(t)
+	w2 = makeMap2(t, types.BehaviorUnspecified)
+	err = w.AbsorbAll(w2)
 	assert.Error(t, err)
```
Comment (koba1t) [inline api/resmap/reswrangler_test.go:909]: If you can, Could you summarize it in one line?

File: api/resource/factory.go [hunk 7]
```diff
@@ -42,7 +42,11 @@ func (rf *Factory) Hasher() ifc.KustHasher {
 
 // FromMap returns a new instance of Resource.
 func (rf *Factory) FromMap(m map[string]interface{}) *Resource {
-	return rf.FromMapAndOption(m, nil)
+	res, err := rf.FromMapAndOption(m, nil)
+	if err != nil {
+		log.Fatal("Option must not be null")
```
Comment (koba1t) [inline api/resource/factory.go:47]: It looks like this line abandoned a non-nil error.
Could you make a change to print the substance of the returned error?

File: api/resource/factory.go [hunk 8]
```diff
@@ -52,19 +56,21 @@ func (rf *Factory) FromMapWithName(n string, m map[string]interface{}) *Resource
 
 // FromMapWithNamespaceAndName returns a new instance with the given "original" namespace.
 func (rf *Factory) FromMapWithNamespaceAndName(ns string, n string, m map[string]interface{}) *Resource {
-	r := rf.FromMapAndOption(m, nil)
+	r, err := rf.FromMapAndOption(m, nil)
+	if err != nil {
+		log.Fatal("Option must not be null")
```
Comment (koba1t) [inline api/resource/factory.go:61]: same here

**Review by koba1t — CHANGES_REQUESTED** [review]
I think we should avoid the `fatal` function.
But we are currently using `log.Fatal()`, and we can fix it with the next PR.
https://github.com/kubernetes-sigs/kustomize/pull/5381/files#diff-9041e805f414eb549e8ebcf4f2b82c1f40403e3ebe86a1f67f9c76718bd91797L65

I think this PR is moving forward to the first step and is useful for kustomize.

@chansuke 
Thanks for your contribution!
I added any comments, Could you check it?

**Comment by chansuke** [comment]
@koba1t I appreciate your feedback and review! I will go fix them.

**Comment by chansuke** [comment]
@koba1t I've addressed your comments and made the necessary changes. Please take a look :pray:
And I'd like to work on this in the next PR if it's possible :bow:

> I think we should avoid the fatal function.
But we are currently using log.Fatal(), and we can fix it with the next PR.

**Comment by koba1t** [comment]
@chansuke 

Thanks for the fix!
Looks good for me!

/lgtm

@natasha41575 
Would you like to review this PR, or do you have any opinions?
In the meantime, I look forward to your reply.

**Comment by natasha41575** [comment]
@koba1t Thanks! I'll take a look on friday

File: api/resource/factory.go [hunk 9]
```diff
@@ -42,7 +42,11 @@ func (rf *Factory) Hasher() ifc.KustHasher {
 
 // FromMap returns a new instance of Resource.
 func (rf *Factory) FromMap(m map[string]interface{}) *Resource {
-	return rf.FromMapAndOption(m, nil)
+	res, err := rf.FromMapAndOption(m, nil)
+	if err != nil {
+		log.Fatalf("failed to create resource from map: %v", err)
```
Comment (natasha41575) [inline api/resource/factory.go:47]: I think the idea of the TODO was to return the error all the way to the top, e.g. completely avoid the use of `log.Fatalf` for this. That would mean `FromMap` and `FromMapWithNamespaceAndName` should the error, instead of doing `log.Fatalf` here, and likewise all the way up.

Looks like you and @koba1t already had a discussion about potentially doing that in a follow-up PR. I think that's fine to do in a followup, but in that case could we keep the TODO here? 

```	
if err != nil {
		// TODO: return err instead of log.
		log.Fatalf("failed to create resource from map: %v", err)
	}
```

File: api/resource/factory.go [hunk 10]
```diff
@@ -52,19 +56,21 @@ func (rf *Factory) FromMapWithName(n string, m map[string]interface{}) *Resource
 
 // FromMapWithNamespaceAndName returns a new instance with the given "original" namespace.
 func (rf *Factory) FromMapWithNamespaceAndName(ns string, n string, m map[string]interface{}) *Resource {
-	r := rf.FromMapAndOption(m, nil)
+	r, err := rf.FromMapAndOption(m, nil)
+	if err != nil {
+		log.Fatalf("failed to create resource from map: %v", err)
```
Comment (natasha41575) [inline api/resource/factory.go:61]: same here

**Review by natasha41575 — CHANGES_REQUESTED** [review]

File: api/resource/factory.go (hunk 9 above)
Comment (chansuke) [reply api/resource/factory.go:47]: @natasha41575 Thank you for the review. Yes, I will put the TODO comment again

**Comment by chansuke** [comment]
@natasha41575 Updated the code. Please take a look.

**Review by natasha41575 — COMMENTED** [review]
/lgtm
/approve

**Review by natasha41575 — APPROVED** [review]
